### PR TITLE
OCPCLOUD-2749: testutils: resourcebuilder: extend fields

### PR DESCRIPTION
**Pull request by damdo**
extend fields of resourcebuilder to allow for more flexibility for the following:

- testutils: resourcebuilder: machinev1beta1.AWSProviderSpec extend fields
- testutils: resourcebuilder: machinev1beta1.Machine extend fields
- testutils: resourcebuilder: machinev1beta1.MachineSet extend fields

**Testing**:

To prove this doesn't break existing use cases I ran cluster-control-plane-machine-set-operator unit tests successfully.
```
Ginkgo ran 11 suites in 5m30.502007958s
Test Suite Passed
```

### CODE REVIEW

File: testutils/resourcebuilder/coalesce.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
+/*
+Copyright 2022 Red Hat, Inc.
```
Comment (JoelSpeed) [inline testutils/resourcebuilder/coalesce.go:2]: Gone back in time? 👀

File: testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go [hunk 2]
```diff
@@ -20,48 +20,90 @@ import (
...
-			},
-		},
-		subnet: machinev1beta1.AWSResourceReference{
+var (
+	defaultInstanceType          = "m6i.xlarge"
+	defaultAvailabilityZone      = "us-east-1a"
+	defaultRegion                = "us-east-1"
+	defaultUserDataSecretName    = "aws-user-data-12345678"
+	defaultCredentialsSecretName = "aws-cloud-credentials"
+	defaultAMIID                 = ptr.To[string]("aws-ami-12345678")
```
Comment (JoelSpeed) [inline testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go:36]: Surprised this is a pointer, does it need to be, for the coalesce to work correctly?

File: testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go [hunk 3]
```diff
@@ -127,26 +154,76 @@ func (m AWSProviderSpecBuilder) BuildRawExtension() *runtime.RawExtension {
...
+// WithAMI sets the AMI for the AWS machine config builder.
+func (m AWSProviderSpecBuilder) WithAMI(ami *machinev1beta1.AWSResourceReference) AWSProviderSpecBuilder {
+	m.ami = ami
+	return m
+}
+
 // WithAvailabilityZone sets the availabilityZone for the AWS machine config builder.
-func (m AWSProviderSpecBuilder) WithAvailabilityZone(az string) AWSProviderSpecBuilder {
+func (m AWSProviderSpecBuilder) WithAvailabilityZone(az *string) AWSProviderSpecBuilder {
 	m.availabilityZone = az
```
Comment (JoelSpeed) [inline testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go:165]: Why not take the address of AZ rather than changing the signature?

File: testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go [hunk 4]
```diff
@@ -127,26 +154,76 @@ func (m AWSProviderSpecBuilder) BuildRawExtension() *runtime.RawExtension {
...
 	return m
 }
 
+// WithBlockDevices sets the BlockDevices for the AWS machine config builder.
+// Use WithBlockDevices(nil), to set them to the zero value.
+func (m AWSProviderSpecBuilder) WithBlockDevices(blockDevices *[]machinev1beta1.BlockDeviceMappingSpec) AWSProviderSpecBuilder {
```
Comment (JoelSpeed) [inline testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go:171]: Make this not a pointer, and take the address of blcokDevices?

File: testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go (hunk 3 above)
Comment (damdo) [reply testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go]: Yes good point! Updated. LMK if the change sounds reasonable and I can extend more fields like so.

File: testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go [hunk 5]
```diff
@@ -20,48 +20,94 @@ import (
...
+	return AWSProviderSpecBuilder{
+		availabilityZone: &defaultAvailabilityZone,
+		instanceType:     &defaultInstanceType,
+		securityGroups:   &defaultSecurityGroups,
+		subnet:           &defaultSubnet,
+	}
 }
 
 // AWSProviderSpecBuilder is used to build out a AWS machine config object.
 type AWSProviderSpecBuilder struct {
```
Comment (JoelSpeed) [inline testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go:98]: Might just want a godoc to explain why these are all pointers/double pointers

File: testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go [hunk 6]
```diff
@@ -71,44 +117,25 @@ func (m AWSProviderSpecBuilder) Build() *machinev1beta1.AWSMachineProviderConfig
...
-		Placement: machinev1beta1.Placement{
-			Region:           "us-east-1",
-			AvailabilityZone: m.availabilityZone,
-		},
-		SecurityGroups: m.securityGroups,
-		Subnet:         m.subnet,
-		UserDataSecret: &corev1.LocalObjectReference{
-			Name: "aws-user-data-12345678",
-		},
+		AMI: CoalesceAWSResourceReference(m.ami, machinev1beta1.AWSResourceReference{
```
Comment (JoelSpeed) [inline testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go:120]: The generic Coalesce you wrote doesn't work here?

File: testutils/resourcebuilder/machine/v1beta1/aws_provider_spec_test.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
+package v1beta1
+
+import (
+	"testing"
+
+	machinev1beta1 "github.com/openshift/api/machine/v1beta1"
+	"k8s.io/utils/ptr"
+)
+
+// ------------- DNM: FOR DEMONSTRATION ONLY ----------------
```
Comment (JoelSpeed) [inline testutils/resourcebuilder/machine/v1beta1/aws_provider_spec_test.go:10]: Why DNM? Having tests isn't harmful is it?

File: testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go [hunk 8]
```diff
@@ -127,26 +154,76 @@ func (m AWSProviderSpecBuilder) BuildRawExtension() *runtime.RawExtension {
...
 
 // WithSubnet sets the subnet for the AWS machine config builder.
 func (m AWSProviderSpecBuilder) WithSubnet(subnet machinev1beta1.AWSResourceReference) AWSProviderSpecBuilder {
-	m.subnet = subnet
+	m.subnet = &subnet
+	return m
+}
+
+// WithUserDataSecret sets the UserDataSecret for the AWS machine config builder.
+func (m AWSProviderSpecBuilder) WithUserDataSecret(userDataSecret *corev1.LocalObjectReference) AWSProviderSpecBuilder {
```
Comment (JoelSpeed) [inline testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go:226]: Can you add a test that shows what happens when I pass `WithUserDataSecret(nil)`?

File: testutils/resourcebuilder/machine/v1beta1/coalesce.go
Comment (JoelSpeed) [inline testutils/resourcebuilder/machine/v1beta1/coalesce.go:1]: Not sure if these are needed, vs using the generic one?

File: testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go (hunk 6 above)
Comment (damdo) [reply testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go:120]: Unfortunately not. That generic Coalesce relies on the provided type being comparable by generics standards.
Slice, map, and function types are not comparable, and so structs that contains them.

so they yield errors like so:
```
[]"github.com/openshift/api/machine/v1beta1".BlockDeviceMappingSpec does not satisfy comparable compiler[InvalidTypeArg](https://pkg.go.dev/golang.org/x/tools/internal/typesinternal#InvalidTypeArg)
```

Thus I've added some custom Coalesce functions for the types that are not adhering to comparable.

---

Another, more flexible option is to go with `interface{}` rather than generics, WDYT?:

```
// Coalesce returns the first value (value) provided it is a non-zero value for the type,
// otherwise it returns a the second value (defaultValue).
func Coalesce(i interface{}, defaultValue interface{}) interface{} {
	// Use reflection to get the underlying type and value
	val := reflect.ValueOf(i)

	// Check if the interface{} is a pointer
	if val.Kind() != reflect.Ptr {
		panic("Coalesce only accepts pointers to a value")
	}

	// Check if the pointer is nil
	if val.IsNil() {
		return defaultValue
	}

	// Dereference the pointer and return the value it points to
	return val.Elem().Interface()
}
```

And then use it like so:
```
LoadBalancers:      resourcebuilder.Coalesce(m.loadBalancers, defaultLoadBalancers).([]machinev1beta1.LoadBalancerReference),
Placement: resourcebuilder.Coalesce(m.placement, machinev1beta1.Placement{
	Region:           resourcebuilder.Coalesce(m.region, defaultRegion).(string),
	AvailabilityZone: resourcebuilder.Coalesce(m.availabilityZone, defaultAvailabilityZone).(string),
}).(machinev1beta1.Placement),
```

File: testutils/resourcebuilder/machine/v1beta1/aws_provider_spec_test.go (hunk 7 above)
Comment (damdo) [reply testutils/resourcebuilder/machine/v1beta1/aws_provider_spec_test.go:10]: I can remove the comment if we are happy with them. Or do we want them to use ginkgo?

File: testutils/resourcebuilder/machine/v1beta1/coalesce.go
Comment (damdo) [reply testutils/resourcebuilder/machine/v1beta1/coalesce.go:1]: They are at the moment, see: https://github.com/openshift/cluster-api-actuator-pkg/pull/331#discussion_r1781367355

File: testutils/resourcebuilder/machine/v1beta1/aws_provider_spec_test.go (hunk 7 above)
Comment (JoelSpeed) [reply testutils/resourcebuilder/machine/v1beta1/aws_provider_spec_test.go:10]: Would prefer ginkgo if you have the time to convert them across?

File: testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go (hunk 6 above)
Comment (JoelSpeed) [reply testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go:120]: No strong opinion either way to be honest, happy to leave as is, but the internal coalesce should be private right?

File: testutils/resourcebuilder/machine/v1beta1/coalesce.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	machinev1beta1 "github.com/openshift/api/machine/v1beta1"
+)
+
+func CoalesceAWSResourceReference(v1 *machinev1beta1.AWSResourceReference, v2 machinev1beta1.AWSResourceReference) machinev1beta1.AWSResourceReference {
```
Comment (JoelSpeed) [inline testutils/resourcebuilder/machine/v1beta1/coalesce.go:23]: Doesn't need to be exported does it?

File: testutils/resourcebuilder/machine/v1beta1/aws_provider_spec_test.go (hunk 7 above)
Comment (damdo) [reply testutils/resourcebuilder/machine/v1beta1/aws_provider_spec_test.go]: Done

File: testutils/resourcebuilder/machine/v1beta1/coalesce.go (hunk 9 above)
Comment (damdo) [reply testutils/resourcebuilder/machine/v1beta1/coalesce.go]: Yeah. Unexported now.

File: testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go (hunk 6 above)
Comment (damdo) [reply testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go]: Ok, went with generics.

**Review by JoelSpeed — COMMENTED** [review]
Looking good, do we want to complete the rest of the fields here or follow up for those in a separate PR, the pattern is looking good and the tests are helpful, thanks

File: testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go [hunk 10]
```diff
@@ -20,48 +20,100 @@ import (
...
+	}
+)
+
+// AWSProviderSpec creates a new AWS machine config builder.
+func AWSProviderSpec() AWSProviderSpecBuilder {
+	return AWSProviderSpecBuilder{
+		availabilityZone: &defaultAvailabilityZone,
+		instanceType:     &defaultInstanceType,
+		securityGroups:   &defaultSecurityGroups,
+		subnet:           &defaultSubnet,
```
[the remark this answers is not in the document]
Comment (damdo) [reply testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go]: I wasn't sure about this, but I checked the main user of this pkg (CPMSO) and we don't need to 👍 So I removed them.

File: testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go [hunk 11]
```diff
@@ -127,26 +169,124 @@ func (m AWSProviderSpecBuilder) BuildRawExtension() *runtime.RawExtension {
...
+
+// WithBlockDevices sets the BlockDevices for the AWS machine config builder.
+// Use WithBlockDevices(nil), to set them to the zero value.
+func (m AWSProviderSpecBuilder) WithBlockDevices(blockDevices []machinev1beta1.BlockDeviceMappingSpec) AWSProviderSpecBuilder {
+	m.blockDevices = &blockDevices
+	return m
+}
+
+// WithCredentialsSecret sets the DeviceIndex for the AWS machine config builder.
+func (m AWSProviderSpecBuilder) WithDeviceIndex(deviceIndex int64) AWSProviderSpecBuilder {
```
Comment (JoelSpeed) [inline testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go:192]: Can we make sure that all of the functions are alphabetically ordered, and have the correct godoc, this one doesn't for example

File: testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go [hunk 12]
```diff
@@ -127,26 +169,124 @@ func (m AWSProviderSpecBuilder) BuildRawExtension() *runtime.RawExtension {
...
+}
+
+// WithMetadataServiceOptions sets the MetadataServiceOptions for the AWS machine config builder.
+func (m AWSProviderSpecBuilder) WithMetadataServiceOptions(opts machinev1beta1.MetadataServiceOptions) AWSProviderSpecBuilder {
+	m.metadataServiceOptions = &opts
+	return m
+}
+
+// WithSpotMarketOptions sets the SpotMarketOptions for the AWS machine config builder.
+func (m AWSProviderSpecBuilder) WithSpotMarketOptions(opts *machinev1beta1.SpotMarketOptions) AWSProviderSpecBuilder {
```
Comment (JoelSpeed) [inline testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go:235]: Should come later, after SecurityGroups?

Comment (damdo) [reply testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go]: TY

File: testutils/resourcebuilder/machine/v1beta1/machine.go [hunk 13]
```diff
@@ -30,47 +37,69 @@ func Machine() MachineBuilder {
...
+	generateName          string
+	labels                map[string]string
+	lifecycleHooks        *machinev1beta1.LifecycleHooks
+	machineSpec           *machinev1beta1.MachineSpec
+	machineSpecObjectMeta *machinev1beta1.ObjectMeta
+	name                  string
+	namespace             string
+	providerID            **string
+	providerSpecBuilder   *resourcebuilder.RawExtensionBuilder
+	providerSpec          *machinev1beta1.ProviderSpec
```
Comment (JoelSpeed) [inline testutils/resourcebuilder/machine/v1beta1/machine.go:53]: Do we have a builder function for this one?

File: testutils/resourcebuilder/machine/v1beta1/machine.go [hunk 14]
```diff
@@ -30,47 +37,69 @@ func Machine() MachineBuilder {
...
-	providerSpecBuilder resourcebuilder.RawExtensionBuilder
-	providerID          *string
+	authoritativeAPI      *machinev1beta1.MachineAuthority
+	annotations           map[string]string
+	creationTimestamp     metav1.Time
+	deletionTimestamp     *metav1.Time
+	generateName          string
+	labels                map[string]string
+	lifecycleHooks        *machinev1beta1.LifecycleHooks
+	machineSpec           *machinev1beta1.MachineSpec
```
Comment (JoelSpeed) [inline testutils/resourcebuilder/machine/v1beta1/machine.go:47]: Do we have a builder function for this?

File: testutils/resourcebuilder/machine/v1beta1/machine.go [hunk 15]
```diff
@@ -30,47 +37,69 @@ func Machine() MachineBuilder {
...
+	machineSpecObjectMeta *machinev1beta1.ObjectMeta
+	name                  string
+	namespace             string
+	providerID            **string
+	providerSpecBuilder   *resourcebuilder.RawExtensionBuilder
+	providerSpec          *machinev1beta1.ProviderSpec
+	taints                []corev1.Taint
```
Comment (JoelSpeed) [inline testutils/resourcebuilder/machine/v1beta1/machine.go:54]: Do we have a builder function for this?

**Comment by damdo** [comment]
All 3 commits now pushed for AWSProvider, Machine and MachineSet resourcebuilder extensions.

@JoelSpeed this is ready for a final review

File: testutils/resourcebuilder/machine/v1beta1/machineset.go [hunk 16]
```diff
@@ -88,32 +142,123 @@ func (m MachineSetBuilder) AsWorker() MachineSetBuilder {
 		WithLabel(machineSetMachineTypeLabelName, "worker")
 }
 
+// WithAnnotations sets the annotations for the machineSet on the machineSet builder.
+// DEPRECATED: use WithMachineSetAnnotations() instead.
```
Comment (JoelSpeed) [inline testutils/resourcebuilder/machine/v1beta1/machineset.go:146]: Why is this special and prefixing the labels/annotations with MachineSet? I wouldn't expect that on other resources.

Do we do something similar on CPMS?

File: testutils/resourcebuilder/machine/v1beta1/machineset.go [hunk 17]
```diff
@@ -27,55 +28,108 @@ const (
...
 
+var (
+	defaultAvailableReplicas      = int32(0)
+	defaultDeletePolicy           = ""
+	defaultFullyLabeledReplicas   = int32(0)
+	defaultMinReadySeconds        = int32(0)
+	defaultObservedGeneration     = int64(0)
+	defaultReadyReplicas          = int32(0)
+	defaultReplicasStatus         = int32(0)
+	defaultSynchronizedGeneration = int64(0)
```
Comment (JoelSpeed) [inline testutils/resourcebuilder/machine/v1beta1/machineset.go:39]: Do we really need default values when they are the zero value for the field?

File: testutils/resourcebuilder/machine/v1beta1/machineset.go [hunk 18]
```diff
@@ -27,55 +28,108 @@ const (
...
 		},
 		Spec: machinev1beta1.MachineSetSpec{
-			Replicas: &m.replicas,
+			AuthoritativeAPI: resourcebuilder.Coalesce(m.authoritativeAPI, defaultAuthoritativeAPI),
+			DeletePolicy:     resourcebuilder.Coalesce(m.deletePolicy, defaultDeletePolicy),
+			MinReadySeconds:  resourcebuilder.Coalesce(m.minReadySeconds, defaultMinReadySeconds),
+			Replicas:         m.replicas,
 			Selector: metav1.LabelSelector{
-				MatchLabels: m.labels,
+				MatchLabels: m.machineTemplateLabels,
```
Comment (JoelSpeed) [inline testutils/resourcebuilder/machine/v1beta1/machineset.go:100]: Is there an occasion you can think of where we might want to try an alternative selector?

File: testutils/resourcebuilder/machine/v1beta1/machineset.go (hunk 16 above)
Comment (damdo) [reply testutils/resourcebuilder/machine/v1beta1/machineset.go:146]: We do have MachineSet (top level annotations/labels on the resource), and we then have the machineTemplate ones, which are then propagated to the children Machines. I wanted to have a way to differentiate the two, as in real MachineSet they could be different right?

File: testutils/resourcebuilder/machine/v1beta1/machineset.go (hunk 18 above)
Comment (damdo) [reply testutils/resourcebuilder/machine/v1beta1/machineset.go:100]: The selector should be matchng the labels in the Template.ObjectMeta.Labels below, so this and that one should match.
Although the MachineSet.ObjectMeta.Labels, are different right?

Comment (JoelSpeed) [reply testutils/resourcebuilder/machine/v1beta1/machineset.go:89]: Yes, MachineSet.ObjectMeta.Labels are different.

My thinking here was that perhaps we might want to use this builder to provide a case where the selector uses match expressions, or, doesn't match the labels on the template.

I think having a builder that allows providing an entire selector, and then defaulting to what you've included here may be a more flexible solution

File: testutils/resourcebuilder/machine/v1beta1/machineset.go (hunk 16 above)
Comment (JoelSpeed) [reply testutils/resourcebuilder/machine/v1beta1/machineset.go:135]: I agree that these should be different, but wondered why we don't just call these top level ones `WithLabels` and `WithAnnotations`, since that's what we do on other objects isn't it?

The ones within the spec I'm happy to be named differently, but thought we should be consistent on the top level object meta ones

Comment (damdo) [reply testutils/resourcebuilder/machine/v1beta1/machineset.go]: Yeah I did that to avoid confusion, but I agree having consistency is also desireable.  In that case we will need to set both labels to the same for AsWorker() now.

Addressed that in the latest commit.

Also testcases that currently use MachineSet.WithLabels() will need updating as now  WithLabels() will only update ms.Labels and not also ms.Spec.Selector.MatchLabels. (e.g. https://github.com/search?q=repo%3Aopenshift%2Fmachine-api-operator%20WithLabel&type=code )

File: testutils/resourcebuilder/machine/v1beta1/machineset.go (hunk 18 above)
Comment (damdo) [reply testutils/resourcebuilder/machine/v1beta1/machineset.go:90]: Makes sense. Extended the builder to accept a custom selector with `WithMachineSetSpecSelector()`, this may then differ from the machineSet.Spec.Template.ObjectMeta.Labels, which can separately be set with `WithMachineTemplateLabels()`.

File: testutils/resourcebuilder/machine/v1beta1/machineset.go [hunk 19]
```diff
@@ -34,48 +35,91 @@ func MachineSet() MachineSetBuilder {
...
-	creationTimestamp   metav1.Time
-	providerSpecBuilder resourcebuilder.RawExtensionBuilder
+	annotations                map[string]string
+	authoritativeAPI           machinev1beta1.MachineAuthority
+	creationTimestamp          metav1.Time
+	deletePolicy               string
+	deletionTimestamp          *metav1.Time
+	generateName               string
+	lifecycleHooks             machinev1beta1.LifecycleHooks
+	machineSetLabels           map[string]string
```
Comment (JoelSpeed) [inline testutils/resourcebuilder/machine/v1beta1/machineset.go:45]: Inconsistent with annotations?

**Review by JoelSpeed — COMMENTED** [review]
/lgtm
/approve
